File: live/AbletonMind/handlers/push.py
"""
Handlers Push 2/3 (Phase 6 start — ADR-0008).

Ableton Push usa Sysex MIDI para LEDs/pads. Frame: F0 00 21 1D 01 01 <cmd> <args> F7.

Bridge envia via `ctrl._send_midi(bytes)`. Em headless / sem Push, devolve
`-32000` com `detected: false`.

Buttons canônicos (subset Cycle 10):
  Play, Record, Stop, Tap Tempo, Metronome, Undo, Mute, Solo, Master,
  Note, Session, Browse, Add Track, Add Device, Setup, Shift, Select.
"""
from ..errors import LIVE_NOT_RUNNING, OBJECT_NOT_FOUND, OUT_OF_RANGE, RpcError
from ..schemas import PushSetButtonLedInput, PushSetModeInput, PushSetPadColorInput
from ._base import Handler, register

# Sysex prefix (Ableton vendor 00 21 1D, Push device family 01 01).
_SYSEX_PREFIX = bytes([0xF0, 0x00, 0x21, 0x1D, 0x01, 0x01])
_SYSEX_END = bytes([0xF7])

# Commands (Push 2/3 — referência docs Ableton Push API).
_CMD_SET_PAD_COLOR = 0x04
_CMD_SET_BUTTON_LED = 0x05
_CMD_SET_MODE = 0x0A

_PUSH_MODES = {"note": 0, "session": 1, "drum": 2, "step": 3}

# Modes para botões.
_BUTTON_MODES = {"solid": 0, "blink": 1, "pulse": 2}

# Map button name → MIDI CC number (canônico Push 2/3 — subset).
_BUTTON_CCS = {
    "Play": 85,
    "Record": 86,
    "Stop": 29,
    "Tap Tempo": 3,
    "Metronome": 9,
    "Undo": 119,
    "Mute": 60,
    "Solo": 61,
    "Master": 28,
    "Note": 50,
    "Session": 51,
    "Browse": 111,
    "Add Track": 53,
    "Add Device": 52,
    "Setup": 30,
    "Shift": 49,
    "Select": 48,
}


def _has_push(ctrl) -> bool:
    """Detecção heurística: ctrl.application().control_surfaces tem alguma com
    name contendo 'Push'? Em headless retorna False."""
    app = getattr(ctrl, "application", None) or getattr(ctrl, "_application", None)
    if app is None:
        return False
    surfaces = list(getattr(app, "control_surfaces", []) or [])
    for s in surfaces:
        if "push" in str(getattr(s, "name", "")).lower():
            return True
    return False


def _send_sysex(ctrl, payload: bytes) -> None:
    sender = getattr(ctrl, "_send_midi", None) or getattr(ctrl, "send_midi", None)
    if sender is None:
        raise RpcError(
            LIVE_NOT_RUNNING,
            "no midi sender on ctrl",
            {"hint": "headless mode"},
        )
    msg = _SYSEX_PREFIX + payload + _SYSEX_END
    sender(tuple(msg))


@register("push.set_pad_color")
class PushSetPadColorHandler(Handler):
    INPUT = PushSetPadColorInput

    def execute(self, params: PushSetPadColorInput) -> dict:
        if not _has_push(self.ctrl):
            raise RpcError(
                LIVE_NOT_RUNNING,
                "Push not detected",
                {"detected": False, "hint": "Connect Push 2 or 3 + activate Control Surface."},
            )
        if params.pad < 0 or params.pad > 63:
            raise RpcError(
                OUT_OF_RANGE,
                "pad out of range",
                {"min": 0, "max": 63, "got": params.pad},
            )
        if params.color < 0 or params.color > 127:
            raise RpcError(
                OUT_OF_RANGE,
                "color out of range",
                {"min": 0, "max": 127, "got": params.color},
            )
        # Sysex: cmd, pad, color
        _send_sysex(self.ctrl, bytes([_CMD_SET_PAD_COLOR, params.pad, params.color]))
        return {"pad": params.pad, "color": params.color, "sent": True}


@register("push.set_button_led")
class PushSetButtonLedHandler(Handler):
    INPUT = PushSetButtonLedInput

    def execute(self, params: PushSetButtonLedInput) -> dict:
        if not _has_push(self.ctrl):
            raise RpcError(
                LIVE_NOT_RUNNING,
                "Push not detected",
                {"detected": False},
            )
        if params.button not in _BUTTON_CCS:
            raise RpcError(
                OBJECT_NOT_FOUND,
                "unknown button",
                {"valid": sorted(_BUTTON_CCS.keys()), "got": params.button},
            )
        if params.color < 0 or params.color > 127:
            raise RpcError(
                OUT_OF_RANGE,
                "color out of range",
                {"min": 0, "max": 127, "got": params.color},
            )
        mode = _BUTTON_MODES.get(params.mode)
        if mode is None:
            raise RpcError(
                OBJECT_NOT_FOUND,
                "unknown mode",
                {"valid": sorted(_BUTTON_MODES.keys()), "got": params.mode},
            )
        cc = _BUTTON_CCS[params.button]
        _send_sysex(self.ctrl, bytes([_CMD_SET_BUTTON_LED, cc, params.color, mode]))
        return {
            "button": params.button,
            "color": params.color,
            "mode": params.mode,
            "sent": True,
        }


@register("push.set_mode")
class PushSetModeHandler(Handler):
    INPUT = PushSetModeInput

    def execute(self, params: PushSetModeInput) -> dict:
        if not _has_push(self.ctrl):
            raise RpcError(LIVE_NOT_RUNNING, "Push not detected", {"detected": False})
        mode_byte = _PUSH_MODES.get(params.mode)
        if mode_byte is None:
            raise RpcError(
                OBJECT_NOT_FOUND,
                "unknown push mode",
                {"valid": sorted(_PUSH_MODES.keys()), "got": params.mode},
            )
        _send_sysex(self.ctrl, bytes([_CMD_SET_MODE, mode_byte]))
        return {"mode": params.mode, "sent": True}

File: src/tools/push.ts
/**
 * Tools Phase 6 — Push 2/3 LED/pad control (ADR-0008).
 *
 * Bridge handler sends MIDI Sysex. In headless or without Push detected,
 * `-32000` with `detected: false`.
 */

import { z } from "zod";

import { defineTool } from "../server/define-tool.js";

const PUSH_BUTTONS = [
  "Play",
  "Record",
  "Stop",
  "Tap Tempo",
  "Metronome",
  "Undo",
  "Mute",
  "Solo",
  "Master",
  "Note",
  "Session",
  "Browse",
  "Add Track",
  "Add Device",
  "Setup",
  "Shift",
  "Select",
] as const;

// ----- push_set_pad_color ----------------------------------------------------

const padInputSchema = z.object({
  pad: z.number().int().min(0).max(63).describe("Push pad grid 0..63 (8x8)."),
  color: z.number().int().min(0).max(127),
});

const padOutputSchema = z.object({
  ok: z.literal(true),
  verified: z.literal(true),
  pad: z.number().int().nonnegative(),
  color: z.number().int().nonnegative(),
  sent: z.literal(true),
});

export const pushSetPadColorTool = defineTool({
  name: "push_set_pad_color",
  description:
    "Set the color of one Ableton Push 2/3 pad on the 0..63 grid. Use for visual feedback or performance layouts after hardware detection. Requires Push 2/3; returns sent=true with the pad and color accepted by the bridge.",
  input: padInputSchema,
  output: padOutputSchema,
  handler: async (input, ctx) => {
    const raw = await ctx.bridge.call("push.set_pad_color", {
      pad: input.pad,
      color: input.color,
    });
    const parsed = padOutputSchema.omit({ ok: true, verified: true }).parse(raw);
    return { ok: true as const, verified: true as const, ...parsed };
  },
});

// ----- push_set_button_led ---------------------------------------------------

const buttonInputSchema = z.object({
  button: z.enum(PUSH_BUTTONS),
  color: z.number().int().min(0).max(127),
  mode: z.enum(["solid", "blink", "pulse"]).optional(),
});

const buttonOutputSchema = z.object({
  ok: z.literal(true),
  verified: z.literal(true),
  button: z.string(),
  color: z.number().int().nonnegative(),
  mode: z.string(),
  sent: z.literal(true),
});

// ----- push_set_mode ---------------------------------------------------------

const modeInputSchema = z.object({
  mode: z.enum(["note", "session", "drum", "step"]),
});

const modeOutputSchema = z.object({
  ok: z.literal(true),
  verified: z.literal(true),
  mode: z.string(),
  sent: z.literal(true),
});

export const pushSetModeTool = defineTool({
  name: "push_set_mode",
  description:
    "Switch Ableton Push 2/3 to note, session, drum, or step mode. Use when preparing a performance or editing workflow on Push hardware. Requires Push 2/3; returns sent=true with the mode accepted by the bridge.",
  input: modeInputSchema,
  output: modeOutputSchema,
  handler: async (input, ctx) => {
    const raw = await ctx.bridge.call("push.set_mode", { mode: input.mode });
    const parsed = modeOutputSchema.omit({ ok: true, verified: true }).parse(raw);
    return { ok: true as const, verified: true as const, ...parsed };
  },
});

export const pushSetButtonLedTool = defineTool({
  name: "push_set_button_led",
  description:
    "Set the LED color and mode for a Push transport or utility button. Use for hardware feedback after Push detection. Requires Push 2/3; defaults mode to solid and returns sent=true with the accepted button/color/mode.",
  input: buttonInputSchema,
  output: buttonOutputSchema,
  handler: async (input, ctx) => {
    const raw = await ctx.bridge.call("push.set_button_led", {
      button: input.button,
      color: input.color,
      mode: input.mode ?? "solid",
    });
    const parsed = buttonOutputSchema.omit({ ok: true, verified: true }).parse(raw);
    return { ok: true as const, verified: true as const, ...parsed };
  },
});
